sw/host/opentitanlib/src/transport/hyperdebug/mod.rs

```diff
@@ -0,0 +1,320 @@
+                        // No data arrived for a while, and the last we got was a command
+                        // prompt, this is what we expect when the command has finished
+                        // successfully.
+                        return Ok(())
+                    } else {
+                        return Err(error.into())
+                    }
+                }
+                Err(error) => {
+                    return Err(error.into())
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("USB device did not match")]
+    NoMatch,
+    #[error("Found no HyperDebug USB device")]
+    NoDevice,
+    #[error("Found multiple HyperDebug USB devices, use --serial")]
+    MultipleDevices,
+    #[error("Error communicating with HyperDebug: {0}")]
+    CommunicationError(&'static str),
+    #[error("Encountered non-unicode path")]
+    UnicodePathError,
+}
+
+impl Transport for Hyperdebug {
+    fn capabilities(&self) -> Capabilities {
+        Capabilities::new(Capability::UART | Capability::GPIO | Capability::SPI)
+    }
+
+    // Crate SPI Target instance, or return one from a cache of previously created instances.
+    fn spi(&self, instance: &str) -> Result<Rc<dyn Target>> {
+        match self.spi_names.get(instance) {
+            Some(&idx) => {
+                if let Some(instance) = self.inner.spis.borrow().get(&idx) {
+                    return Ok(Rc::clone(instance));
+                }
+                let instance: Rc<dyn Target> = Rc::new(spi::HyperdebugSpiTarget::open(
+                    &self,
+                    idx,
+                )?);
+                self.inner.spis.borrow_mut().insert(idx, Rc::clone(&instance));
+                return Ok(instance);
+            }
+            _ => Err(TransportError::InvalidInstance("spi", instance.to_string()).into())
+        }
+    }
+    
+    // Crate Uart instance, or return one from a cache of previously created instances.
+    fn uart(&self, instance: &str) -> Result<Rc<dyn Uart>> {
+        match self.uart_ttys.get(instance) {
+            Some(tty) => {
+                if let Some(instance) = self.inner.uarts.borrow().get(tty) {
+                    return Ok(Rc::clone(instance));
+                }
+                let instance: Rc<dyn Uart> = Rc::new(uart::HyperdebugUart::open(
+                            &self,
+                            tty,
+                        )?);
+                self.inner.uarts.borrow_mut().insert(tty.clone(), Rc::clone(&instance));
+                Ok(instance)
+            }
+            _ => Err(TransportError::InvalidInstance("uart", instance.to_string()).into())
+        }
+    }
+    
+    // Crate GpioPin instance, or return one from a cache of previously created instances.
+    fn gpio_pin(&self, pinname: &str) -> Result<Rc<dyn GpioPin>> {
+        Ok(match self.inner.gpio.borrow_mut().entry(pinname.to_string()) {
```

Comment:
I think that there is an `or_insert` function on Entry that is helpful here.

Reply:
I could not figure out how to make that work.  The HyperdebugGpioPin::open() method is expensive, and has side effects, so we do not want to use plain or_insert(), which evaluates its "default" argument in all cases, even when the entry is populated already.  The is an or_insert_with(), which takes a closure (zero argument function) to produce the default value, but I have a Result<>, and I would like to bail out of the insertion and the call altogether, if the return of HyperdebugGpioPin::open() is Err(...), and lacking exception stack unwinding I do not know a way of doing that from inside such a closure, in Rust.

Reply:
Ah. yes, that is unfortunate. I don't believe Rust gives you an out here. =/